File: crates/hyper_rhi/src/vulkan/texture_view.rs
/*
 * Copyright (c) 2024, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
*/

use ash::vk;

use crate::vulkan::GraphicsDevice;

#[derive(Clone, Debug)]
pub(super) struct TextureViewDescriptor {
    pub(super) format: vk::Format,
}

pub(crate) struct TextureView {
    view: vk::ImageView,

    graphics_device: GraphicsDevice,
}

impl TextureView {
    pub(super) fn new(
        graphics_device: &GraphicsDevice,
        image: vk::Image,
        descriptor: &TextureViewDescriptor,
    ) -> Self {
        let component_mapping = vk::ComponentMapping::default()
            .r(vk::ComponentSwizzle::IDENTITY)
            .g(vk::ComponentSwizzle::IDENTITY)
            .b(vk::ComponentSwizzle::IDENTITY)
            .a(vk::ComponentSwizzle::IDENTITY);

        let subresource_range = vk::ImageSubresourceRange::default()
            .aspect_mask(vk::ImageAspectFlags::COLOR)
            .base_mip_level(0)
            .level_count(1)
            .base_array_layer(0)
            .layer_count(1);

        let create_info = vk::ImageViewCreateInfo::default()
            .image(image)
            .view_type(vk::ImageViewType::TYPE_2D)
            .format(descriptor.format)
            .components(component_mapping)
            .subresource_range(subresource_range);

        let view = unsafe {
            graphics_device
                .device()
                .create_image_view(&create_info, None)
        }
        .expect("failed to create image view");

        Self {
            view,

            graphics_device: graphics_device.clone(),
        }
    }
}

impl Drop for TextureView {
    fn drop(&mut self) {
        unsafe {
            self.graphics_device
                .device()
                .destroy_image_view(self.view, None);
        }
    }
}

File: hyper_rhi/src/hyper_rhi/vulkan/texture_view.cpp
/*
 * Copyright (c) 2026-present, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_rhi/vulkan/texture_view.hpp"

#include "hyper_rhi/vulkan/graphics_device.hpp"
#include "hyper_rhi/vulkan/render_conversion.hpp"
#include "hyper_rhi/vulkan/texture.hpp"
#include "hyper_rhi/vulkan/utils.hpp"

namespace he {

VulkanTextureView::VulkanTextureView(VulkanGraphicsDevice &graphics_device, const TextureViewDescriptor &desc)
    : m_graphics_device(graphics_device)
{
    const auto *texture = graphics_device.get_internal_state<VulkanTexture>(desc.texture);

    constexpr auto component_mapping = VkComponentMapping {
        .r = VK_COMPONENT_SWIZZLE_IDENTITY,
        .g = VK_COMPONENT_SWIZZLE_IDENTITY,
        .b = VK_COMPONENT_SWIZZLE_IDENTITY,
        .a = VK_COMPONENT_SWIZZLE_IDENTITY,
    };

    const auto format = desc.texture.format();

    const auto mip_levels = desc.mip_levels.value_or(desc.texture.mip_levels());
    const auto array_layers = [&desc]() -> u32 {
        if (desc.array_layers.has_value()) {
            return desc.array_layers.value();
        }

        switch (desc.dimension) {
        case ViewDimension::D1:
        case ViewDimension::D2:
        case ViewDimension::D3:
            return 1;
        case ViewDimension::D2Array:
            return desc.texture.extent().depth;
        case ViewDimension::Cube:
            return 6;
        default:
            HE_UNREACHABLE();
        }
    }();

    const auto subresource_range = VkImageSubresourceRange {
        .aspectMask = map_aspect(format),
        .baseMipLevel = desc.base_mip_level,
        .levelCount = mip_levels,
        .baseArrayLayer = desc.base_array_layer,
        .layerCount = array_layers,
    };

    const auto image_view_create_info = VkImageViewCreateInfo {
        .sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO,
        .pNext = nullptr,
        .flags = 0,
        .image = texture->raw(),
        .viewType = map_view_dimension(desc.dimension),
        .format = map_format(format),
        .components = component_mapping,
        .subresourceRange = subresource_range,
    };

    HE_VK_CHECK(
        vkCreateImageView(m_graphics_device.device(), &image_view_create_info, nullptr, &m_raw),
        "Failed to create vulkan texture view");
    HE_ASSERT(m_raw != VK_NULL_HANDLE);
}

VulkanTextureView::~VulkanTextureView() { vkDestroyImageView(m_graphics_device.device(), m_raw, nullptr); }

} // namespace he
